horovod/spark/keras/datamodule.py

```diff
@@ -0,0 +1,173 @@
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.has_val and self.val_reader:
+            self.val_reader.stop()
+        if self.train_reader:
+            self.train_reader.stop()
+        super().__exit__(type, value, traceback)
+
+    @contextlib.contextmanager
+    def empty_batch_reader(self):
+        yield None
+
+    def train_data(self):
+        return self.make_dataset(self.train_reader, self.train_batch_size,
+                                 self.is_batch_reader, shuffle=self.shuffle, cache=self.inmemory_cache_all)
+
+    def val_data(self):
+        return self.make_dataset(self.val_reader, self.val_batch_size,
+                                 self.is_batch_reader, shuffle=False, cache=self.inmemory_cache_all) if self.val_reader else None
+
+
+class NVTabularDataModule(DataModule):
+    """NVTabular-based DataModule for KerasEstimator for GPU-accelerated data loading of tabular datasets.
+
+    Note: requires `label_cols`, `categorical_cols`, and `continuous_cols` to be explicitly provided."""
+    def __init__(self, label_cols=[], categorical_cols=[], continuous_cols=[], **kwargs):
+        super().__init__(**kwargs)
+        self.label_cols = label_cols
+        self.categorical_cols = categorical_cols
+        self.continuous_cols = continuous_cols
+
+    @staticmethod
+    def seed_fn():
+        """
+        Generate consistent dataloader shuffle seeds across workers
+        Reseeds each worker's dataloader each epoch to get fresh a shuffle
+        that's consistent across workers.
+        """
+        import horovod.tensorflow.keras as hvd
+        import numpy as np
+        import tensorflow as tf
+        hvd.init()
+        seed = np.random.randint(0, tf.int32.max)
+        seed_tensor = tf.constant(seed)
+        root_seed = hvd.broadcast(seed_tensor, name="shuffle_seed", root_rank=0)
+        return root_seed
+
+    @staticmethod
+    def to_dense(X, labels):
+        """Convert NVTabular's ragged representation of array types to dense arrays."""
+        import tensorflow as tf
+        retX = {}
+        for col, tensor in X.items():
+            if isinstance(tensor, tuple):
+                retX[col] = tf.RaggedTensor.from_row_lengths(values=tensor[0][:,0], row_lengths=tensor[1][:,0]).to_tensor()
+            else:
+                retX[col] = tensor
+        return retX, labels
+
+    def train_data(self):
+        import horovod.tensorflow.keras as hvd
+        from nvtabular.loader.tensorflow import KerasSequenceLoader
+        return KerasSequenceLoader(self.train_dir,
+                                   batch_size=self.train_batch_size,
+                                   label_names=self.label_cols,
+                                   cat_names=self.categorical_cols,
+                                   cont_names=self.continuous_cols,
+                                   engine="parquet",
+                                   shuffle=True,
```

Comment:
shuffle = self.shuffle?

I think there are many cases that users don't want training samples to be shuffled.